### PR TITLE
Add OSTree ITE-4 resolver

### PR DESCRIPTION
**Pull request by adityasaky**
Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

**Fixes issue #**: N/A

**Description of the changes being introduced by the pull request**:

Adds an OSTree specific resolver based off @lukpueh's work in #584.

NOTE: #584 must be merged first.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by adityasaky** [comment]
```bash
$ in-toto-run -x -n test-repo-dir -g -p "ostree:torizoncore/dunfell/colibri"

{
 "signatures": [
  {
   "keyid": "b247995a78cf2c06021f3e729ceff07a7164fce4",
   "other_headers": "04000108001d162104b247995a78cf2c06021f3e729ceff07a7164fce40502644ac2e5",
   "signature": "281cd7d0cd8d066adb8cfd707d762261a363d60b48d0b207332807ac31e8d941781af6094625ac44b32d39aa14fd8c24471968bc9b8a24c184110236fb9218ca631fa4079502ca137cad8b8a0a2cadd47fc702cb697a02dd0497d6fb85782c627386bd90632fdff43b9431666abf77657889b3b8e71e28dba32da33ac92f1bebd46d5b3d30da1d2457633066d0c186a886b2cdd9d25e621a995dfdb6d36b3c12cb9676f0e99e93dcb13480381f0cca0f2dd05afde95a4b3fa16f9eaeafec525d307609553a2eb50e52ef3bcc8716f7079903b7353259bf1381d71fb6fcbe7ada8701ec26a3466482ec593c7565f12e2fd99ddb0aba049edc098efb4f423c30f8dcbc09dc6686e03746753258b46a2fab1e0c25ab9b51a46da7e711479ce1c2b4d695590a0b3b41ca7bbfcd5296c0f961cccda3f933f66feef9eda39bf30ed6dcdf7ca6e5a130c8f391d88cd7abb41b845aee6b74698b81f3116775c42ae535f2ed0af100c8343b81f62bb2725fb7bc2b3e5424976ca8b4a83eca21131ccc20243cd00389a73309db6a2354cb60f91c14ed27c0520ad9794228d82c15febcaf0e8208a809c40c616b3612d0484c0ac48e8b5b3f6ae9deef3184c7f725faeca4fe28d87e84b5e9d9bf07ff3c9e4a242cc9a64dcb48b8e6a5ba40109b645953ac6c34416d21f0318c6d972f50000bace7c89fbd110fb55fbd8d582faa4fc57eda18"
  }
 ],
 "signed": {
  "_type": "link",
  "byproducts": {},
  "command": [],
  "environment": {},
  "materials": {},
  "name": "test-repo-dir",
  "products": {
   "ostree:torizoncore/dunfell/colibri": {
    "sha256": "a00e4b999df9bc9bbc14994f5ef46a16f3e2eef4aad3007f58d6d8f0342842e2"
   }
  }
 }
}
```

Also supports using `--base-path` to specify the repo.

```bash
$ in-toto-run -x -n test-base-path -g -p "ostree:torizoncore/dunfell/colibri" --base-path="repo/"

{
 "signatures": [
  {
   "keyid": "b247995a78cf2c06021f3e729ceff07a7164fce4",
   "other_headers": "04000108001d162104b247995a78cf2c06021f3e729ceff07a7164fce40502644ac2dd",
   "signature": "ccfa72e8d50ba01d71a91e993c7d2ea36cd3406bf268621b6860b8207d5914bda94d914bfb769e25d1d3a83bac8fa6af3f631b5e10aaba1574d961e90b8e45a4d68263e287c778b64a8f0fdc67d42ebb28932d43e37e99c074dd94b9180751f7c241a796dfea1594f03d5b9143015abe3b855ea92f92f61a2c006ae11ec7dfa71ed367e5cbc1e83ace0f3e656c17678d742dbf9ce3590b1decbe340092d06db933ace73c72c4b0118f1aca64fa631cd1c5f1071747487080e1bbe97cb68b881e463613502ca23700022abb97eb540a6ff20374ecaf3c56e37bf45948608b557054b52d16488b2f3a04608b389789256b71d61d1ea9463e64c0c2da6c70fc30d853cd205e368461a900078096b059e81967acd25ef5ec21bf67fcf2eea7c7351d524a5e1eaa657f31d0d3aec228ef56bcc858a7f93bc63e268351a528b865bf93e800bd9a1d08cf7876d8953d6d6e2b411678c6e1c56f9cbf5bd189e525838f16f91b44716fc7ed32047ba2141aebc37354385a1f2b3f56a0cd8d8a91c1e84b8ea4c3ba737052162f9f1bb4a807eab8d5ae88fa8e5d877635eb3dd09ef6362ca487448b7af788e02b3de510bf51a3afa97858a621c81accc0631c87a1bd4a738c2b0a36020b773b22a7dfc95dfc691b30f2cd3d8828cedacb7c792de0c887ec418187149edcedcea90e6fe9a59c91d8527dfa4b002c9debe6644b4dd0f78888db"
  }
 ],
 "signed": {
  "_type": "link",
  "byproducts": {},
  "command": [],
  "environment": {},
  "materials": {},
  "name": "test-base-path",
  "products": {
   "ostree:torizoncore/dunfell/colibri": {
    "sha256": "a00e4b999df9bc9bbc14994f5ef46a16f3e2eef4aad3007f58d6d8f0342842e2"
   }
  }
 }
}
```

File: in_toto/resolver/_resolver.py [hunk 1]
```diff
@@ -205,3 +205,71 @@ def hash_artifacts(self, uris):
...
+        digest = digest_filename(object_path, algorithm=self._HASH_ALGORITHM)
+
+        return {self._HASH_ALGORITHM: digest.hexdigest()}
+
+    def hash_artifacts(self, uris):
+        hashes = {}
+
+        if self._base_path:
+            original_cwd = os.getcwd()
+            # FIXME: Re-raise for backwards-compat; should remove try/except
```
Comment (lukpueh) [inline in_toto/resolver/_resolver.py]: No need to worry about backwards-compat in OSTreeResolver. :) I even changed this for FileResolver now in #592.

File: tests/test_ostree_resolver.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import unittest
+
+from in_toto.exceptions import OSTreeRefNotFound
+from in_toto.resolver import OSTreeResolver
+
+class TestOSTreeResolver(unittest.TestCase):
+    """Test OSTree resolver implementation. """
+
+    def test_hash_for_branch_with_base_path(self):
+        resolver = OSTreeResolver(base_path="tests/resolver/ostree_repo")
```
Comment (lukpueh) [inline tests/test_ostree_resolver.py]: Using paths relative to cwd makes the test suite less flexible. I usually use paths relative to the test module, e.g.:
https://github.com/in-toto/in-toto/blob/fbbe7927257086e832deb8a860392000eb58f489/tests/test_in_toto_match_products.py#L15-L17

File: in_toto/resolver/_resolver.py [hunk 3]
```diff
@@ -205,3 +205,71 @@ def hash_artifacts(self, uris):
...
+        """Helper to add scheme back after recording the hash."""
+
+        return f"{self.SCHEME}:{path}"
+
+    def _hash(self, path):
+        """Helper to hash OSTree commits."""
+
+        ref_path = os.path.join("refs", "heads", path)
+        if not os.path.exists(ref_path):
+            raise OSTreeRefNotFound(f"OSTree ref '{path}' not found")
```
Comment (lukpueh) [inline in_toto/resolver/_resolver.py]: The check here doesn't guarantee that there won't be a `FileNotFoundError` below, or a different error from open (missing permissions etc.) 

Do you anticipate users to handle `OSTreeRefNotFound`? If not, I'd just use the builtins.

**Review by lukpueh — CHANGES_REQUESTED** [review]
I can't judge if this, that is the code in `_hash`, is correct wrt OSTree, but it looks reasonable to me. Please address minor nits, and we can merge as experimental.

File: in_toto/resolver/_resolver.py (hunk 3 above)
Comment (adityasaky) [reply in_toto/resolver/_resolver.py]: Removed the exception altogether.

**Comment by adityasaky** [comment]
Fixed linter.

File: in_toto/resolver/_resolver.py [hunk 4]
```diff
@@ -203,3 +203,67 @@ def hash_artifacts(self, uris):
...
+        digest = digest_filename(object_path, algorithm=self._HASH_ALGORITHM)
+
+        return {self._HASH_ALGORITHM: digest.hexdigest()}
+
+    def hash_artifacts(self, uris):
+        hashes = {}
+
+        if self._base_path:
+            original_cwd = os.getcwd()
+            try:
```
Comment (lukpueh) [inline in_toto/resolver/_resolver.py]: Oh, I meant you can remove the try/except block, and just os.chdir.

**Review by lukpueh — APPROVED** [review]
💯

**Pull request merged**